daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerInteraction.scala

```diff
@@ -850,18 +873,19 @@ class JsonLedgerClient(
       new RuntimeException("setTime is not supported when running DAML Script over the JSON API."))
   }
 
-  // Check that the party in the token matches the given party.
-  private def validateTokenParty(party: Ref.Party, what: String): Future[Unit] = {
-    tokenPayload.party match {
-      case None =>
-        Future.failed(
-          new RuntimeException(
-            s"Tried to $what as ${party} but token does not provide a unique party identifier"))
-      case Some(tokenParty) if (!(tokenParty == party)) =>
-        Future.failed(
-          new RuntimeException(
-            s"Tried to $what as ${party} but token is only valid for $tokenParty"))
-      case _ => Future.unit
+  // Check that the parties in the token are a superset of the given parties.
+  private def validateTokenParties(parties: OneAnd[Set, Ref.Party], what: String): Future[Unit] = {
+    val tokenParties = Set(tokenPayload.readAs ++ tokenPayload.actAs: _*)
+    // First check is just for a nicer error message and would be covered by the second
+    if (tokenParties.isEmpty) {
+      Future.failed(
+        new RuntimeException(
+          s"Tried to $what as ${parties.toList.mkString(" ")} but token contains no parties."))
+    } else if (tokenParties == Set[String](parties.toList: _*)) {
```

Comment:
```suggestion
    } else if (tokenParties == parties.toSet.toSet[String]) {
```

Reply:
Do you really mean `toSet` twice or `toList` followed by `toSet`?

Reply:
Or maybe just `toSet` once?

Reply:
The first `toSet` doesn't take a type parameter; the second one does, which is where you can do the widen that you're doing with natural transformation to `List` and then using covariance in the merged code. The second `toSet` will return the receiver with its type changed.